Add Functional NodeDef Support

kwokcb opened the pull request:
## Implementation

This are the basic changes to handle the current "functional nodedef" proposal: #2355 as an
new way to form associations between a definition and it's nodegraph implementation.
This new variant is purely additive for the data model and code logic and API interface and internal logic.

## Modifications.

- Update node definition dictionary in the document cache to consider `nodedefs` with child `nodegraph`s (implementations). Non-child implementations have higher precedence.   
- Update definition query for implementations / nodegraphs to check for parent nodedef first.
- Add `DefintionOptions` options class to which contains the option to add as child or not. Default is false. Additional arguments can be added here vs a list of args in the previous interface.
- Update C++ `Document::addNodeDefFromGraph()` interface to take  `DefinitionsOptions` as a new optional argument. 
- Add in interface to check for shared implementations in an inheritance hierarchy:
  - `NodeDef::getMatchingDefinitions()` and `NodeDef::hasSharedImplementation()`.
- Add in conversion utility to "inline" implementations. General enough to handle any implementation but has a firewall check for now to only allow "functional nodegraphs". Very easy to remove this to accept any implementation variant.
  - `NodeDef::inlineImplementation()`
- Add in Python and Javascript wrappers.
 
## Tests

- Add a `functional_nodedef.mtlx` test used for unit and codegen / render tests.
- Update unit tests to run definition creation test for both sibling (current) and child (new) options as defined by `DefinitionOptions`.
- Update `Javascript` unit tests. Expected:
   ✔ Build document
    ✔ Create NodeDef from NodeGraph with child implementation
    ✔ Create NodeDef from NodeGraph with referencing implementation
- Add unit tests for inlining, looking for shared implementations and inheritance queries.
- Manual testing in graph editor. 
  - Value modifications
  - Show definition
 
 ## Unit Test  Results
 
 ### Definition Created with Sibling Reference
 <details>
 
 ```xml
 <?xml version="1.0"?>
<materialx version="1.39">
  <nodegraph name="test_colorcorrect">
    <multiply name="AlphaGain" type="float">
      <input name="in1" type="float" nodename="inputAlpha" />
      <input name="in2" type="float" interfacename="AlphaGain_in2" />
    </multiply>
    <add name="AlphaOffset" type="float">
      <input name="in1" type="float" nodename="AlphaGain" />
      <input name="in2" type="float" interfacename="AlphaOffset_in2" />
    </add>
    <multiply name="ColorGain" type="color3">
      <input name="in1" type="color3" nodename="inputColor" />
      <input name="in2" type="color3" interfacename="ColorGain_in2" />
    </multiply>
    <add name="ColorOffset" type="color3">
      <input name="in1" type="color3" nodename="ColorGain" />
      <input name="in2" type="color3" interfacename="ColorOffset_in2" />
    </add>
    <constant name="inputColor" type="color3">
      <input name="value" type="color3" interfacename="inputColor_value" />
    </constant>
    <constant name="inputAlpha" type="float">
      <input name="value" type="float" interfacename="inputAlpha_value" />
    </constant>
    <output name="out" type="color3" nodename="ColorOffset" />
    <output name="out1" type="float" nodename="AlphaOffset" />
    <input name="AlphaGain_in2" type="float" value="0.8" uiname="AlphaGain in2" uifolder="Common" />
    <input name="AlphaOffset_in2" type="float" value="1" uiname="AlphaOffset in2" uifolder="Common" />
    <input name="ColorGain_in2" type="color3" value="0.9, 0.9, 0.9" uiname="ColorGain in2" uifolder="Common" />
    <input name="ColorOffset_in2" type="color3" value="0.379147, 0.0341412, 0.0341412" uiname="ColorOffset in2" uifolder="Common" />
    <input name="inputColor_value" type="color3" value="0.5, 0.5, 0.5" uiname="inputColor value" uifolder="Common" />
    <input name="inputAlpha_value" type="float" value="1" uiname="inputAlpha value" uifolder="Common" />
  </nodegraph>
  <nodedef name="ND_test_colorcorrect" node="test_colorcorrect" version="1.0" isdefaultversion="false" nodegroup="adjustment" uiname="test_colorcorrect Version: 1.0" doc="This is version 1 of the definition for the graph: NG_test_colorcorrect">
    <input name="AlphaGain_in2" type="float" value="0.8" uiname="AlphaGain in2" uifolder="Common" />
    <input name="AlphaOffset_in2" type="float" value="1" uiname="AlphaOffset in2" uifolder="Common" />
    <input name="ColorGain_in2" type="color3" value="0.9, 0.9, 0.9" uiname="ColorGain in2" uifolder="Common" />
    <input name="ColorOffset_in2" type="color3" value="0.379147, 0.0341412, 0.0341412" uiname="ColorOffset in2" uifolder="Common" />
    <input name="inputColor_value" type="color3" value="0.5, 0.5, 0.5" uiname="inputColor value" uifolder="Common" />
    <input name="inputAlpha_value" type="float" value="1" uiname="inputAlpha value" uifolder="Common" />
    <output name="out" type="color3" />
    <output name="out1" type="float" />
  </nodedef>
  <nodegraph name="NG_test_colorcorrect" nodedef="ND_test_colorcorrect">
    <multiply name="AlphaGain" type="float">
      <input name="in1" type="float" nodename="inputAlpha" />
      <input name="in2" type="float" interfacename="AlphaGain_in2" />
    </multiply>
    <add name="AlphaOffset" type="float">
      <input name="in1" type="float" nodename="AlphaGain" />
      <input name="in2" type="float" interfacename="AlphaOffset_in2" />
    </add>
    <multiply name="ColorGain" type="color3">
      <input name="in1" type="color3" nodename="inputColor" />
      <input name="in2" type="color3" interfacename="ColorGain_in2" />
    </multiply>
    <add name="ColorOffset" type="color3">
      <input name="in1" type="color3" nodename="ColorGain" />
      <input name="in2" type="color3" interfacename="ColorOffset_in2" />
    </add>
    <constant name="inputColor" type="color3">
      <input name="value" type="color3" interfacename="inputColor_value" />
    </constant>
    <constant name="inputAlpha" type="float">
      <input name="value" type="float" interfacename="inputAlpha_value" />
    </constant>
    <output name="out" type="color3" nodename="ColorOffset" />
    <output name="out1" type="float" nodename="AlphaOffset" />
  </nodegraph>
</materialx>
```

</details>

### Definition Created with Child 

<details>

```xml
<?xml version="1.0"?>
<materialx version="1.39">
  <nodegraph name="test_colorcorrect">
    <multiply name="AlphaGain" type="float">
      <input name="in1" type="float" nodename="inputAlpha" />
      <input name="in2" type="float" interfacename="AlphaGain_in2" />
    </multiply>
    <add name="AlphaOffset" type="float">
      <input name="in1" type="float" nodename="AlphaGain" />
      <input name="in2" type="float" interfacename="AlphaOffset_in2" />
    </add>
    <multiply name="ColorGain" type="color3">
      <input name="in1" type="color3" nodename="inputColor" />
      <input name="in2" type="color3" interfacename="ColorGain_in2" />
    </multiply>
    <add name="ColorOffset" type="color3">
      <input name="in1" type="color3" nodename="ColorGain" />
      <input name="in2" type="color3" interfacename="ColorOffset_in2" />
    </add>
    <constant name="inputColor" type="color3">
      <input name="value" type="color3" interfacename="inputColor_value" />
    </constant>
    <constant name="inputAlpha" type="float">
      <input name="value" type="float" interfacename="inputAlpha_value" />
    </constant>
    <output name="out" type="color3" nodename="ColorOffset" />
    <output name="out1" type="float" nodename="AlphaOffset" />
    <input name="AlphaGain_in2" type="float" value="0.8" uiname="AlphaGain in2" uifolder="Common" />
    <input name="AlphaOffset_in2" type="float" value="1" uiname="AlphaOffset in2" uifolder="Common" />
    <input name="ColorGain_in2" type="color3" value="0.9, 0.9, 0.9" uiname="ColorGain in2" uifolder="Common" />
    <input name="ColorOffset_in2" type="color3" value="0.379147, 0.0341412, 0.0341412" uiname="ColorOffset in2" uifolder="Common" />
    <input name="inputColor_value" type="color3" value="0.5, 0.5, 0.5" uiname="inputColor value" uifolder="Common" />
    <input name="inputAlpha_value" type="float" value="1" uiname="inputAlpha value" uifolder="Common" />
  </nodegraph>
  <nodedef name="ND_test_colorcorrect" node="test_colorcorrect" version="1.0" isdefaultversion="false" nodegroup="adjustment" uiname="test_colorcorrect Version: 1.0" doc="This is version 1 of the definition for the graph: NG_test_colorcorrect">
    <nodegraph name="NG_test_colorcorrect">
      <multiply name="AlphaGain" type="float">
        <input name="in1" type="float" nodename="inputAlpha" />
        <input name="in2" type="float" interfacename="AlphaGain_in2" />
      </multiply>
      <add name="AlphaOffset" type="float">
        <input name="in1" type="float" nodename="AlphaGain" />
        <input name="in2" type="float" interfacename="AlphaOffset_in2" />
      </add>
      <multiply name="ColorGain" type="color3">
        <input name="in1" type="color3" nodename="inputColor" />
        <input name="in2" type="color3" interfacename="ColorGain_in2" />
      </multiply>
      <add name="ColorOffset" type="color3">
        <input name="in1" type="color3" nodename="ColorGain" />
        <input name="in2" type="color3" interfacename="ColorOffset_in2" />
      </add>
      <constant name="inputColor" type="color3">
        <input name="value" type="color3" interfacename="inputColor_value" />
      </constant>
      <constant name="inputAlpha" type="float">
        <input name="value" type="float" interfacename="inputAlpha_value" />
      </constant>
      <output name="out" type="color3" nodename="ColorOffset" />
      <output name="out1" type="float" nodename="AlphaOffset" />
    </nodegraph>
    <input name="AlphaGain_in2" type="float" value="0.8" uiname="AlphaGain in2" uifolder="Common" />
    <input name="AlphaOffset_in2" type="float" value="1" uiname="AlphaOffset in2" uifolder="Common" />
    <input name="ColorGain_in2" type="color3" value="0.9, 0.9, 0.9" uiname="ColorGain in2" uifolder="Common" />
    <input name="ColorOffset_in2" type="color3" value="0.379147, 0.0341412, 0.0341412" uiname="ColorOffset in2" uifolder="Common" />
    <input name="inputColor_value" type="color3" value="0.5, 0.5, 0.5" uiname="inputColor value" uifolder="Common" />
    <input name="inputAlpha_value" type="float" value="1" uiname="inputAlpha value" uifolder="Common" />
    <output name="out" type="color3" />
    <output name="out1" type="float" />
  </nodedef>
</materialx>
```

</details>
 
### Graph Editor : Std Library Definition Using Child Nodegraph

https://github.com/user-attachments/assets/f0c70ea9-af13-4e7d-96d7-5fcf65fa34b8

### Graph Editor: Created Definition Using Child Nodegraph

- Instance
<img width="2008" height="1362" alt="image" src="https://github.com/user-attachments/assets/7a8e904d-a5c4-437a-937f-8c09a88b2967" />

- Definition Expanded
<img width="1940" height="770" alt="image" src="https://github.com/user-attachments/assets/a4afdc73-8af1-48d0-a2c7-fde90d960599" />

kwokcb committed:
Initial test.

kwokcb commented on source/MaterialXCore/Document.cpp:
```diff
@@ -99,17 +108,27 @@ class Document::Cache
+            // So append them to the end of the implementation list assocaited 
+            // with any existing nodedef entry (or create a new one if does not exist).
+            //
+            for (const auto& [nodedefKey, appendImplementations] : nonFuncNodeDefMap)
```
Search order is strictly set here for definitions: functional defs first, all others after. Usage priority is decided elsewhere but generally it's a "first found" logic (per target). 

For this initial check-in, we only look for child nodegraphs. Child implementations could be added if / when desired.

kwokcb replied:
This was a remaining question, and I think your logic makes sense. It also addresses a previous concern from Doug about how to extend implementation associations. 

- If there is no overlap then order does not matter. 
- If there is then we can give non-embedded nodegraph implementation a higher priority by adding them to the implementation list first (so that they are found first when looking for implementations).

Note that this does no affect being able to get all the implementations back.

kwokcb commented on source/MaterialXCore/Node.cpp:
```diff
@@ -719,7 +719,14 @@ void NodeGraph::modifyInterfaceName(const string& inputPath, const string& inter
 NodeDefPtr NodeGraph::getNodeDef() const
 {
-    NodeDefPtr nodedef = resolveNameReference<NodeDef>(getNodeDefString());
+    ElementPtr parent = getSelfNonConst()->getParent();
```
"Backwards" search looks for parent definitions now.

kwokcb commented on source/MaterialXCore/Definition.cpp:
```diff
@@ -156,6 +156,12 @@ void Implementation::setNodeDef(ConstNodeDefPtr nodeDef)
 
 NodeDefPtr Implementation::getNodeDef() const
 {
+    ElementPtr parent = getSelfNonConst()->getParent();
```
"Backwards" search looks for parent definitions now.

kwokcb commented on libraries/stdlib/stdlib_defs.mtlx:
```diff
@@ -2148,56 +2148,263 @@
     <input name="in2" type="float" value="1.0" />
     <output name="out" type="float" defaultinput="in1" />
+
+    <nodegraph name="NG_safepower_float">
```
Any / or all graphs could reside here. Both new and old mechanism still work.

kwokcb replied:
I will revert this to not break for anyone parsing this directly.

kwokcb committed:
Merge branch 'AcademySoftwareFoundation:main' into functional_nodedefs

kwokcb committed:
Merge branch 'AcademySoftwareFoundation:main' into functional_nodedefs

kwokcb committed:
Merge branch 'AcademySoftwareFoundation:main' into functional_nodedefs

kwokcb committed:
Merge branch 'AcademySoftwareFoundation:main' into functional_nodedefs

kwokcb committed:
Merge branch 'AcademySoftwareFoundation:main' into functional_nodedefs

kwokcb committed:
Merge branch 'AcademySoftwareFoundation:main' into functional_nodedefs

kwokcb committed:
Merge branch 'AcademySoftwareFoundation:main' into functional_nodedefs

kwokcb committed:
Merge branch 'AcademySoftwareFoundation:main' into functional_nodedefs

kwokcb committed:
Add create options and unit test.

kwokcb committed:
Add doc for DefinitionsOptions arg on addNodeDefFromGraph().

kwokcb committed:
Merge branch 'AcademySoftwareFoundation:main' into functional_nodedefs

kwokcb committed:
Merge branch 'AcademySoftwareFoundation:main' into functional_nodedefs

kwokcb committed:
Merge branch 'AcademySoftwareFoundation:main' into functional_nodedefs

kwokcb committed:
Merge branch 'AcademySoftwareFoundation:main' into functional_nodedefs

kwokcb committed:
Merge branch 'AcademySoftwareFoundation:main' into functional_nodedefs

ld-kerley reviewed (commented):
I think this looks good to me.

+1 to reverting the data library change for now - until we have socialized this change and validated its not going to inconvenience anyone if we start to update the data library.

I left one outstanding question about order of precedence - but maybe Ive got the wrong end of the stick....

kwokcb committed:
Merge branch 'AcademySoftwareFoundation:main' into functional_nodedefs

kwokcb committed:
Review updates

- Update precedence logic so that non-functional implementations have higher priority.
-  Revert std lib sample changes.
- Add in functional_nodedef.mtlx test file
   - Tests for implementation search of child nodegraph
   - Tests for precedence by adding a non-functional implementation graph afterwards.

kwokcb commented on source/MaterialXTest/MaterialXCore/Node.cpp:
```diff
@@ -675,7 +677,64 @@ TEST_CASE("Organization", "[nodegraph]")
+        REQUIRE(referenceNodeGraph != nullptr);
+        if (referenceNodeGraph)
+        {
+            referenceNodeGraph->setNodeDefString(nodeDefName);
```
Superscede the child nodegraph impl with one that is not a child. As all changes cause a nodedef->impl cache rebuild this allows for this dynamic behaviour.

kwokcb reviewed (commented):
Review changes made.

kwokcb committed:
Fix Windows test warnings.

Fix warnings as errors in test.

kwokcb committed:
Add Javascript bindings + unit test.

kwokcb committed:
Merge branch 'AcademySoftwareFoundation:main' into functional_nodedefs

kwokcb committed:
Merge branch 'AcademySoftwareFoundation:main' into functional_nodedefs

kwokcb committed:
Merge branch 'AcademySoftwareFoundation:main' into functional_nodedefs

kwokcb committed:
Merge branch 'AcademySoftwareFoundation:main' into functional_nodedefs

kwokcb committed:
Merge branch 'main' into functional_nodedefs

kwokcb committed:
Merge branch 'AcademySoftwareFoundation:main' into functional_nodedefs

kwokcb committed:
Merge branch 'AcademySoftwareFoundation:main' into functional_nodedefs

kwokcb committed:
Merge branch 'AcademySoftwareFoundation:main' into functional_nodedefs

kwokcb committed:
Merge branch 'AcademySoftwareFoundation:main' into functional_nodedefs

kwokcb committed:
Merge branch 'main' into functional_nodedefs

kwokcb committed:
Merge remote-tracking branch 'upstream/main' into functional_nodedefs

kwokcb committed:
Add inlining, inheritance, and shared implementation logic  and more extensive testing.

kwokcb committed:
Merge remote-tracking branch 'upstream/main' into functional_nodedefs

kwokcb committed:
Auto sync.

kwokcb committed:
Merge remote-tracking branch 'upstream/main' into functional_nodedefs

kwokcb committed:
Fix merge.

kwokcb committed:
Merge remote-tracking branch 'aswf/main' into functional_nodedefs

kwokcb committed:
Merge remote-tracking branch 'upstream/main' into functional_nodedefs

kwokcb committed:
Merge branch 'AcademySoftwareFoundation:main' into functional_nodedefs

kwokcb committed:
Merge branch 'AcademySoftwareFoundation:main' into functional_nodedefs

kwokcb committed:
Merge remote-tracking branch 'upstream/main' into functional_nodedefs

kwokcb committed:
Merge branch 'AcademySoftwareFoundation:main' into functional_nodedefs

kwokcb committed:
Merge branch 'AcademySoftwareFoundation:main' into functional_nodedefs

kwokcb committed:
Merge remote-tracking branch 'upstream/main' into functional_nodedefs

kwokcb committed:
Merge branch 'AcademySoftwareFoundation:main' into functional_nodedefs

kwokcb committed:
Merge remote-tracking branch 'aswf/main' into functional_nodedefs

kwokcb committed:
Fix playwright js test.

kwokcb committed:
Port remaning test code to playwright for js tests.

kwokcb committed:
Merge branch 'AcademySoftwareFoundation:main' into functional_nodedefs
